Invalid trigger for unused_closure_parameter when parameter is used in string interpolation

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
~> swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files in current working directory
# ... non problematic files
/.../Mapper.swift:53:31: warning: Unused Closure Parameter Violation: Unused parameter "input" in a closure should be replaced with _. (unused_closure_parameter)
/.../Mapper.swift:53:38: warning: Unused Closure Parameter Violation: Unused parameter "context" in a closure should be replaced with _. (unused_closure_parameter)
/.../Mapper.swift:53:47: warning: Unused Closure Parameter Violation: Unused parameter "output" in a closure should be replaced with _. (unused_closure_parameter)
Done linting! Found 3 violations, 0 serious in 37 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
```bash
~> swiftlint version
0.25.0
````
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
`Homebrew`
* Paste your configuration file:

```yml
included:
  - Sources

opt_in_rules:
  - array_init
  - attributes
  - closure_end_indentation
  - closure_spacing
  - conditional_returns_on_newline
  - contains_over_first_not_nil
  - empty_count
  - explicit_enum_raw_value
  - explicit_init
  - explicit_top_level_acl
  - explicit_type_interface
  - extension_access_modifier
  - fatal_error_message
  - first_where
  - force_unwrapping
  - implicitly_unwrapped_optional
  - joined_default_parameter
  - let_var_whitespace
  - literal_expression_end_indentation
  - multiline_arguments
  - multiline_parameters
  - nimble_operator
  - no_extension_access_modifier
  - no_grouping_extension
  - number_separator
  - object_literal
  - operator_usage_whitespace
  - overridden_super_call
  - pattern_matching_keywords
  - private_outlet
  - prohibited_super_call
  - quick_discouraged_call
  - redundant_nil_coalescing
  - single_test_class
  - sorted_imports
  - strict_fileprivate
  - switch_case_on_newline
  - trailing_closure
  - unneeded_parentheses_in_closure_argument
  - vertical_parameter_alignment_on_call

disabled_rules:
  - extension_access_modifier
  - explicit_enum_raw_value
  - void_return
  - explicit_type_interface
  - number_separator
  - trailing_whitespace
  - function_body_length
  - line_length
  - redundant_optional_initialization
  - explicit_top_level_acl
  - identifier_name
  - no_grouping_extension

nesting:
  type_level: 2

colon:
  apply_to_dictionaries: false

statement_position:
  statement_mode: uncuddled_else
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcode-select -p`)?
```bash
~> xcode-select -p
/Applications/Xcode.app/Contents/Developer
```

As the title state, the rule doesn't appear to see the variables referenced via string interpolation in the scope of this function.
<img width="1017" alt="screen shot 2018-02-23 at 5 24 52 pm" src="https://user-images.githubusercontent.com/159331/36620135-a46ffa52-18bf-11e8-8d3e-6a81932411fd.png">
<img width="1015" alt="screen shot 2018-02-23 at 5 25 02 pm" src="https://user-images.githubusercontent.com/159331/36620136-a480a6c2-18bf-11e8-8f22-96e188c953ca.png">

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -119,6 +119,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#2212](https://github.com/realm/SwiftLint/issues/2212)
 
+* Fix a false positive in `unused_closure_parameter` rule when a parameter
+  is used in a string interpolation.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#2062](https://github.com/realm/SwiftLint/issues/2062)
+
 ## 0.25.1: Lid Locked
 
 This is the last release to support building with Swift 3.2 and Swift 3.3.

--- a/Rules.md
+++ b/Rules.md
@@ -18465,6 +18465,12 @@ hoge(arg: num) { num in
 })(FileManager.default)
 ```
 
+```swift
+withPostSideEffect { input in
+    if true { print("\(input)") }
+}
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>

--- a/Source/SwiftLintFramework/Extensions/SyntaxMap+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/SyntaxMap+SwiftLint.swift
@@ -12,18 +12,11 @@ extension SyntaxMap {
                 .intersects(byteRange)
         }
 
-        func notIntersect(_ token: SyntaxToken) -> Bool {
-            return !intersect(token)
-        }
-
         guard let startIndex = tokens.index(where: intersect) else {
             return []
         }
         let tokensBeginningIntersect = tokens.lazy.suffix(from: startIndex)
-        if let endIndex = tokensBeginningIntersect.index(where: notIntersect) {
-            return Array(tokensBeginningIntersect.prefix(upTo: endIndex))
-        }
-        return Array(tokensBeginningIntersect)
+        return Array(tokensBeginningIntersect.filter(intersect))
     }
 
     internal func kinds(inByteRange byteRange: NSRange) -> [SyntaxKind] {

--- a/Source/SwiftLintFramework/Rules/UnusedClosureParameterRule.swift
+++ b/Source/SwiftLintFramework/Rules/UnusedClosureParameterRule.swift
@@ -38,6 +38,11 @@ public struct UnusedClosureParameterRule: ASTRule, ConfigurationProviderRule, Co
             ({ (manager: FileManager) in
               print(manager)
             })(FileManager.default)
+            """,
+            """
+            withPostSideEffect { input in
+                if true { print("\\(input)") }
+            }
             """
         ],
         triggeringExamples: [
@@ -130,15 +135,18 @@ public struct UnusedClosureParameterRule: ASTRule, ConfigurationProviderRule, Co
                                                                   length: range.length),
                     // if it's the parameter declaration itself, we should skip
                     byteRange.location > paramOffset,
-                    case let tokens = file.syntaxMap.tokens(inByteRange: byteRange),
-                    // a parameter usage should be only one token
-                    tokens.count == 1 else {
-                    continue
+                    case let tokens = file.syntaxMap.tokens(inByteRange: byteRange) else {
+                        continue
                 }
 
+                let token = tokens.first(where: { token -> Bool in
+                    return SyntaxKind(rawValue: token.type) == .identifier &&
+                        token.offset == byteRange.location &&
+                        token.length == byteRange.length
+                })
+
                 // found a usage, there's no violation!
-                if let token = tokens.first, SyntaxKind(rawValue: token.type) == .identifier,
-                    token.offset == byteRange.location, token.length == byteRange.length {
+                guard token == nil else {
                     return nil
                 }
             }

